src/main/java/org/candlepin/model/PoolCurator.java

```diff
@@ -942,97 +901,44 @@ public List<Pool> getBySubscriptionIds(String ownerId, Collection<String> subIds
-        for (Entry<String, Entitlement> entry : subIdMap.entrySet()) {
-            SimpleExpression subscriptionExpr = Restrictions.eq("sourceSub.subscriptionId", entry.getKey());
-            Criterion srcEntIsNull = Restrictions.isNull("sourceEntitlement");
-            Criterion srcEntEquals = Restrictions.eq("sourceEntitlement", entry.getValue());
-
-            Junction subscriptionJunction = Restrictions.and(subscriptionExpr)
-                .add(Restrictions.or(srcEntIsNull, srcEntEquals));
-
-            subIdMapCriteria.add(subscriptionJunction);
-        }
-
-        return currentSession()
-            .createCriteria(Pool.class)
-            .createAlias("sourceSubscription", "sourceSub")
-            .add(Restrictions.eq("owner.id", ownerId))
-            .add(Restrictions.ge("quantity", 0L))
-            .add(Restrictions.gtProperty("consumed", "quantity"))
-            .add(Restrictions.or(subIdMapCriteria.toArray(exampleCriteria))).list();
-    }
-
-    @Transactional
-    public Pool replicate(Pool pool) {
-        pool.setSourceEntitlement(null);
-
-        this.currentSession().replicate(pool, ReplicationMode.EXCEPTION);
+        List<Pool> result = new ArrayList<>();
 
-        return pool;
-    }
-
-    private static final String CONSUMER_FILTER = "Entitlement_CONSUMER_FILTER";
+        String jpql = """
+            SELECT p FROM Pool p
+            WHERE p.owner.id = :ownerId
+            AND p.quantity >= 0
+            AND p.consumed > p.quantity
+            AND p.sourceSubscription.subscriptionId IN :subscriptionIds
+            AND (p.sourceEntitlement IS NULL OR p.sourceEntitlement IN :entitlements)
+            """;
 
-    @SuppressWarnings("checkstyle:indentation")
-    public int getNoOfDependentEntitlements(String entitlementId) {
-        Filter consumerFilter = disableConsumerFilter();
-        Integer result = (Integer) currentSession()
-            .createCriteria(Entitlement.class)
-            .setProjection(Projections.rowCount())
-            .createCriteria("pool")
-            .createCriteria("sourceEntitlement")
-            .add(Restrictions.idEq(entitlementId))
-            .list()
-            .get(0);
-
-        enableIfPrevEnabled(consumerFilter);
-        return result;
-    }
+        TypedQuery<Pool> query = getEntityManager().createQuery(jpql, Pool.class)
+            .setParameter("ownerId", ownerId);
 
-    // TODO: watch out for performance. Should we limit the certificates
-    // retrieved?
-    @SuppressWarnings("unchecked")
-    public List<EntitlementCertificate> retrieveEntCertsOfPoolsWithSourceEntitlement(String entId) {
-        return currentSession().createCriteria(EntitlementCertificate.class)
-            .createCriteria("entitlement")
-            .createCriteria("pool")
-            .createCriteria("sourceEntitlement")
-            .add(Restrictions.idEq(entId))
-            .list();
-    }
+        for (Map<String, Entitlement> block: this.partitionMap(subIdMap)) {
```

Comment:
Space between the colon and block variable

Reply:
Additionally, we need to partition the map differently. Since we have three params, two of which are blocks, our partition size needs to be the lesser of the in-block size or the parameter limit minus one, divided by two.

Reply:
done